crates/ruff_python_semantic/src/red_knot/types/infer.rs

```diff
@@ -0,0 +1,980 @@
+        })
+    }
+
+    fn typing_context(&self) -> TypingContext {
+        TypingContext::local(self.db, self.scope, &self.types)
+    }
+
+    fn local_definition_ty(&mut self, symbol: ScopeSymbolId) -> DefinitionType {
+        let symbol = self.symbol_table.symbol(symbol);
+        let mut definitions = symbol
+            .definitions()
+            .iter()
+            .filter_map(|definition| self.definition_tys.get(definition).copied());
+
+        let Some(first) = definitions.next() else {
+            return DefinitionType::Type(Type::Unbound);
+        };
+
+        if let Some(second) = definitions.next() {
+            let context = self.typing_context();
+            let mut builder = UnionTypeBuilder::new(&context);
+            builder = builder.variant(first).variant(second);
+
+            for variant in definitions {
+                builder = builder.variant(variant);
+            }
+
+            // The fact that the interner is local to a body means that we can't reuse the same union type
+            // across different call sites. But that's something we aren't doing yet anyway. Our interner doesn't
+            // deduplicate union types that are identical.
+            DefinitionType::Union {
+                ty: builder.build(),
+            }
+        } else {
+            DefinitionType::Type(first)
+        }
+    }
+}
+
+enum DefinitionType {
+    Union { ty: UnionType },
+    Type(Type),
+}
+
+impl DefinitionType {
+    fn into_type(self, types: &mut TypeInferenceBuilder) -> Type {
+        match self {
+            DefinitionType::Union { ty } => types.union_ty(ty),
+            DefinitionType::Type(ty) => ty,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use ruff_db::file_system::FileSystemPathBuf;
+    use ruff_db::vfs::system_path_to_file;
+
+    use crate::db::tests::TestDb;
+    use crate::module::resolver::{set_module_resolution_settings, ModuleResolutionSettings};
+    use crate::name::Name;
+    use crate::red_knot::types::{public_symbol_ty_by_name, Type, TypingContext};
+
+    fn setup_db() -> TestDb {
+        let mut db = TestDb::new();
+
+        set_module_resolution_settings(
+            &mut db,
+            ModuleResolutionSettings {
+                extra_paths: Vec::new(),
+                workspace_root: FileSystemPathBuf::from("/src"),
+                site_packages: None,
+                custom_typeshed: None,
+            },
+        );
+
+        db
+    }
+
+    fn assert_public_ty(db: &TestDb, file_name: &str, symbol_name: &str, expected: &str) {
+        let file = system_path_to_file(db, file_name).expect("Expected file to exist.");
+
+        let ty = public_symbol_ty_by_name(db, file, symbol_name).unwrap_or(Type::Unknown);
+        assert_eq!(ty.display(&TypingContext::global(db)).to_string(), expected);
+    }
+
+    #[test]
+    fn follow_import_to_class() -> anyhow::Result<()> {
+        let db = setup_db();
+
+        db.memory_file_system().write_files([
```

Comment:
Can `TestDb` have a wrapper method so we don't have to always write `memory_file_system()` here? It seems redundant.

Reply:
Possibly. I don't want to change this as part of this PR